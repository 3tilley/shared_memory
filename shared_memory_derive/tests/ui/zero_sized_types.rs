//! Zero sized types are NOT allowed

use shared_memory::SharedMemCast;

#[derive(SharedMemCast)]
pub enum EmptyEnum {}

#[derive(SharedMemCast)]
pub struct Empty;

#[derive(SharedMemCast)]
pub struct Empty2 {
}

#[derive(SharedMemCast)]
pub struct Empty3();

#[derive(SharedMemCast)]
pub struct StillEmpty {
    x: (),
}

#[derive(SharedMemCast)]
pub struct FieldEmpty {
    y: i32,
    x: (),
    z: f64,
    w: [i32; 0],
}

fn main() {}
