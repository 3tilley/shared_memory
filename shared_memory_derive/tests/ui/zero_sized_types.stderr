error: Empty enums types cannot be casted from shared memory
 --> $DIR/zero_sized_types.rs:6:10
  |
6 | pub enum EmptyEnum {}
  |          ^^^^^^^^^

error: Zero-sized types cannot be casted from shared memory
 --> $DIR/zero_sized_types.rs:9:12
  |
9 | pub struct Empty;
  |            ^^^^^

error: Zero-sized types cannot be casted from shared memory
  --> $DIR/zero_sized_types.rs:12:12
   |
12 | pub struct Empty2 {
   |            ^^^^^^

error: Zero-sized types cannot be casted from shared memory
  --> $DIR/zero_sized_types.rs:16:12
   |
16 | pub struct Empty3();
   |            ^^^^^^

error[E0277]: the trait bound `(): shared_memory::cast::SharedMemCast` is not satisfied
  --> $DIR/zero_sized_types.rs:20:8
   |
20 |     x: (),
   |        ^^ the trait `shared_memory::cast::SharedMemCast` is not implemented for `()`
   |
   = note: required by `shared_memory::cast::AssertIsSharedMemCast`

error[E0277]: the trait bound `(): shared_memory::cast::SharedMemCast` is not satisfied
  --> $DIR/zero_sized_types.rs:26:8
   |
26 |     x: (),
   |        ^^ the trait `shared_memory::cast::SharedMemCast` is not implemented for `()`
   |
   = note: required by `shared_memory::cast::AssertIsSharedMemCast`

error[E0277]: the trait bound `[i32; 0]: shared_memory::cast::SharedMemCast` is not satisfied
  --> $DIR/zero_sized_types.rs:28:8
   |
28 |     w: [i32; 0],
   |        ^^^^^^^^ the trait `shared_memory::cast::SharedMemCast` is not implemented for `[i32; 0]`
   |
   = help: the following implementations were found:
             <[T; 1024] as shared_memory::cast::SharedMemCast>
             <[T; 1048576] as shared_memory::cast::SharedMemCast>
             <[T; 1073741824] as shared_memory::cast::SharedMemCast>
             <[T; 10] as shared_memory::cast::SharedMemCast>
           and 55 others
   = note: required by `shared_memory::cast::AssertIsSharedMemCast`

For more information about this error, try `rustc --explain E0277`.
